libtenzir/builtins/operators/select.cpp

```diff
@@ -77,10 +77,29 @@ class select_operator final
     return "select";
   }
 
-  auto optimize(expression const& filter, event_order order) const
+  auto optimize(expression const& filter, event_order order,
+                select_optimization const& selection) const
     -> optimize_result override {
-    (void)filter;
-    return optimize_result::order_invariant(*this, order);
+    if (config_.fields.empty() && selection.fields_of_interest.empty()) {
+      return optimize_result::order_invariant(*this, order);
+    }
+    if (!selection.fields_of_interest.empty()) {
+      auto config_fields = config_.fields;
+      auto selection_fields = selection.fields_of_interest;
+      std::sort(config_fields.begin(), config_fields.end());
+      std::sort(selection_fields.begin(), selection_fields.end());
+      std::vector<std::string> intersection;
+      std::set_intersection(config_fields.begin(), config_fields.end(),
+                            selection_fields.begin(), selection_fields.end(),
+                            std::back_inserter(intersection));
```

Comment:
The intersection works incorrectly when using type extractors. Consider this example:

```
…
| select :ip
| select src_ip
```

You cannot know the type of the field `src_ip` ahead of time. The intersection is only empty when `src_ip` does not exist in the input event, or there are no IP type fields, or when `src_ip` is not of type IP.

I think we need to back to the drawing board to exactly figure out whether we can optimize this at all for TQL1.

Reply:
Considering the select | select logic needs more discussion, should I not combine selections for now (stop the later selections from getting pushed up).